File: src/com/pailsom/leetcode/easy/RomanToInt.java
package com.pailsom.leetcode.easy;

import java.util.HashMap;

public class RomanToInt {
    public static void main(String[] args) {
        System.out.println(romanToInt("IVII"));
    }

    private static Integer romanToInt(String s) {
        HashMap<Character,Integer> h=new HashMap<>();
        h.put('I',1);
        h.put('X',10);
        h.put('V',5);
        h.put('C',100);
        h.put('D',500);
        h.put('M',1000);
        h.put('L',50);

        Character prev='k';
        int ans=0;
        for(int i=0;i<s.length();i++){
            Character q=s.charAt(i);
            ans+=h.get(q);
            if(prev=='I' && (q=='X' || q=='V'))
                ans=ans-2;
            else if(prev=='X' && (q=='L' || q=='C'))
                ans=ans-20;
            else if(prev=='C' && (q=='D' || q=='M'))
                ans=ans-200;
            prev=q;
        }
        return ans;
    }
}

File: src/com/kotlin/leetcode/easy/RomanToInt.kt
package com.kotlin.leetcode.easy

fun main() {
    println(romanToInt("LVIII"))
}

fun romanToInt(s: String): Int {
    val map = hashMapOf(
        'I' to 1,
        'V' to 5,
        'X' to 10,
        'L' to 50,
        'C' to 100,
        'D' to 500,
        'M' to 1000
    )
    var ans = 0
    var pre = 'k'
    s.toCharArray().forEach { c ->
        ans+=map[c]!!
        if(pre=='I' && (c=='V' || c=='X')) ans-=2
        if(pre=='X' && (c=='L' || c=='C')) ans-=20
        if(pre=='C' && (c=='D' || c=='M')) ans-=200
        pre = c
    }
    return ans
}